File: MattEland.Emergence.Domain/Obstacle.fs
module MattEland.Emergence.Domain.Obstacles

open MattEland.Emergence.Domain

type ObstacleType =
    | Wall = 0
    | Column = 1
    | Service = 2
    | Data = 3
    | ThreadPool = 4
    | Barrier = 5
    
let getMaxHealth obstacleType =
    match obstacleType with
        | ObstacleType.Wall -> 10        
        | _ -> 5
        
type Obstacle (position: Position, obstacleType: ObstacleType) = 
  inherit DestructibleObject(position, getMaxHealth obstacleType)
  member this.ObstacleType = obstacleType

  override this.AsciiCharacter =
    match obstacleType with
      | ObstacleType.Barrier -> '-'
      | ObstacleType.Column -> 'o'
      | ObstacleType.Service -> '*'
      | ObstacleType.Data -> 'd'
      | ObstacleType.ThreadPool -> '~'
      | _ -> '#'

  override this.ZIndex = 50

  interface IInteractive with
    member this.interact source =
      seq {
        yield new DisplayMessage("You can't go that way. Moron.")
      }

File: MattEland.Emergence.Model/Entities/Obstacle.cs
using System;
using System.Collections.Generic;
using MattEland.Emergence.Model.Messages;

namespace MattEland.Emergence.Model.Entities
{
    public class Obstacle : WorldObject, IInteractive
    {
        public ObstacleType ObstacleType { get; }

        public Obstacle(Position pos, ObstacleType obstacleType) : base(pos, Guid.NewGuid())
        {
            ObstacleType = obstacleType;
        }

        public override string ForegroundColor
        {
            get
            {
                switch (ObstacleType)
                {
                    case ObstacleType.Wall:
                        return GameColors.SlateBlue;
                    case ObstacleType.Service:
                        return GameColors.Orange;
                    case ObstacleType.Data:
                        return GameColors.Purple;
                    case ObstacleType.ThreadPool:
                        return GameColors.Blue;
                    default:
                        return GameColors.DarkGray;
                }                
            }
        }

        public override char AsciiChar
        {
            get
            {
                switch (ObstacleType)
                {
                    case ObstacleType.Wall:
                        return '#';
                    case ObstacleType.Column:
                        return 'O';
                    case ObstacleType.Service:
                        return '*';
                    case ObstacleType.Data:
                        return 'd';
                    case ObstacleType.ThreadPool:
                        return '~';
                    case ObstacleType.Barrier:
                        return 'X';
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }
        }

        public override int ZIndex => 30;
        public IEnumerable<GameMessage> Interact(Actor actor)
        {
            yield return new DisplayTextMessage("You can't go that way.");
        }
    }
}